Userland/Libraries/LibCore/IOSurface.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2024, Aliaksandr Kalenik <kalenik.aliaksandr@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/OwnPtr.h>
+#include <LibCore/IOSurface.h>
+
+#if !defined(AK_OS_MACOS)
+static_assert(false, "This file must only be used for macOS");
+#endif
+
+#define FixedPoint FixedPointMacOS
+#define Duration DurationMacOS
+#include <IOSurface/IOSurface.h>
+#undef FixedPoint
+#undef Duration
+
+namespace Core {
+
+#define kIOSurfacePixelFormatBGRA8888 'BGRA'
+
+struct IOSurfaceHandle::IOSurfaceRefWrapper {
+    IOSurfaceRef ref;
+};
+
+IOSurfaceHandle::IOSurfaceHandle(OwnPtr<IOSurfaceRefWrapper>&& ref_wrapper)
+    : m_ref_wrapper(move(ref_wrapper))
+{
+}
+
+IOSurfaceHandle::IOSurfaceHandle(IOSurfaceHandle&& other) = default;
+IOSurfaceHandle& IOSurfaceHandle::operator=(IOSurfaceHandle&& other) = default;
+
+IOSurfaceHandle::~IOSurfaceHandle()
+{
+    if (m_ref_wrapper)
+        CFRelease(m_ref_wrapper->ref);
+}
+
+IOSurfaceHandle IOSurfaceHandle::create(size_t width, size_t height)
+{
+    size_t bytes_per_element = 4;
+    uint32_t pixel_format = kIOSurfacePixelFormatBGRA8888;
+
+    CFNumberRef width_number = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &width);
+    CFNumberRef height_number = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &height);
+    CFNumberRef bytes_per_element_number = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &bytes_per_element);
+    CFNumberRef pixel_format_number = CFNumberCreate(kCFAllocatorDefault, kCFNumberSInt32Type, &pixel_format);
+
+    CFMutableDictionaryRef props = CFDictionaryCreateMutable(kCFAllocatorDefault,
+        0,
+        &kCFTypeDictionaryKeyCallBacks,
+        &kCFTypeDictionaryValueCallBacks);
+
+    CFDictionarySetValue(props, kIOSurfaceWidth, width_number);
+    CFDictionarySetValue(props, kIOSurfaceHeight, height_number);
+    CFDictionarySetValue(props, kIOSurfaceBytesPerElement, bytes_per_element_number);
+    CFDictionarySetValue(props, kIOSurfacePixelFormat, pixel_format_number);
+
+    auto* ref = IOSurfaceCreate(props);
+
+    CFRelease(width_number);
+    CFRelease(height_number);
+    CFRelease(bytes_per_element_number);
+    CFRelease(pixel_format_number);
+    CFRelease(props);
+
+    return IOSurfaceHandle(make<IOSurfaceRefWrapper>(ref));
+}
+
+MachPort IOSurfaceHandle::create_mach_port() const
+{
+    auto port = IOSurfaceCreateMachPort(m_ref_wrapper->ref);
+    return MachPort::adopt_right(port, MachPort::PortRight::SendOnce);
+}
+
+IOSurfaceHandle IOSurfaceHandle::from_mach_port(MachPort port)
```

Comment:
```suggestion
IOSurfaceHandle IOSurfaceHandle::from_mach_port(MachPort const& port)
```
maybe? I feel uneasy about it unnecessarily copying the port right into the argument. It probably doesn't if you pass it through `move()`, but still better express that explicitly.

Reply:
I changed this to `MachPort&&`, so the port is destroyed, but you get the IOSurface in return.